### PR TITLE
Snappy: Add support for snapcraft builds

### PR DESCRIPTION
**Pull request by kinnison**
This commit enables building snaps for the various platforms supported by snapcraft.io

Currently the snap is under my personal page, and despite building it here, we'll keep that until after all-hands where I'll have a chat with t-release people about what we might do for an official snap.  So far I've had no joy looking at official methods of distributing rustup in other channels such as flatpak.

### CODE REVIEW

File: .travis.yml
Comment (tesuji) [inline .travis.yml]: Use `if [ "$TRAVIS_BRANCH" != "stable" ]`

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: I would prefer to rename this to `mkdir -p snapbuild/snap` to avoid repeatedly name.

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: This variable appears to be unused.

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: Double quote `"$TARGET"` to avoid word splitting.

**Review by tesuji — COMMENTED** [review]
Honestly, I may know less about snapcraft than you. I would prefer to 
make `ci/snapcraft.sh` a bash script with `.bash` extension instead to
make use of bash features - array, also because this script is not intended
to used by other parties (just in our CI).

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: We may want to use bash feature - array instead.

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: Try `! commmand -v snapcraft > /dev/null`.

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: Try `[ -z ... ]` instead.

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: This too.

File: .travis.yml
Comment (kinnison) [reply .travis.yml]: Doh!  Done.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: I think I picked `snap/snap` because of how snapcraft uses `snap` but of course it's not necessary at the top level, so I'll fix that.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Yep, well spotted.  I've fixed that.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Done.  Thanks

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: I'm not a shell person.  I'm happy if you can tell me what to do there, otherwise this can stay until you want to fix it later

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Fair.  Done.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: I have no words for why I didn't use `-z`  Fixed.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Ditto.

**Comment by kinnison** [comment]
@lzutao Thank you for the comprehensive review.  I believe I've tackled everything you hilighted (except the one thing I don't feel confident in) -- could you recheck please?

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: No need for `2>/dev/null` part.

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: Try `head -n 3 Cargo.toml | grep ^version | cut -d\" -f2`

File: ci/snapcraft.sh
Comment (tesuji) [inline ci/snapcraft.sh]: Do you want to use `$GRADE` here too?

File: ci/snapcraft.sh
Comment (tesuji) [reply ci/snapcraft.sh]: Basic uses of bash array if you want to try: https://devhints.io/bash#arrays

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Okay, dropped.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: I'm not sure I approve of this, in case the version string moves in the header.  Really I ought to be parsing the `.toml` but I think this is moderately more sensible than the `head` approach.  Why are you advocating that instead?

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Yes, well spotted.  I was clearly asleep when making the updates before.

File: ci/snapcraft.sh
Comment (tesuji) [reply ci/snapcraft.sh]: It's just a suggestion. You could skip it.

**Comment by tesuji** [comment]
r=me with or with out using bash array, unless others disagree.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Okay, then for now I think I will, thank you.

**Comment by kinnison** [comment]
Thank you @lzutao 

We won't be merging this until I've spoken to the release team about snaps, but it's good to know you're now satisfied with my awful shell code :-D

**Review by rbtcollins — APPROVED** [review]
Sure, why not.

**Review by ghost — APPROVED** [review]

**Comment by kinnison** [comment]
Now that we're a bit further on in our releases, and Windows builds are no longer *always* failing; I'll give this some more thought.

**Comment by kinnison** [comment]
I've started a discussion with Snap people about github actions for this kind of thing.  Depending on how that goes, I'll either rework this PR, or close it and open a fresh one for continued discussion around `rustup` and snapcraft.

File: ci/actions-templates/linux-builds-template.yaml
Comment (tesuji) [inline ci/actions-templates/linux-builds-template.yaml:173]:
```suggestion
        if: matrix.snap_arch
```
here and there

Comment (kinnison) [reply ci/actions-templates/linux-builds-template.yaml:173]: I prefer the explicit comparison since it's meant to be a string if it's set.

**Comment by kinnison** [comment]
I now need to wait for an appropriate publish action to be written and we can proceed further.

File: ci/actions-templates/linux-builds-template.yaml
Comment (jhenstridge) [inline ci/actions-templates/linux-builds-template.yaml]: I've published a `v1` tag for this action.  I'd suggest using that instead of `master`.

File: ci/snapcraft.sh
Comment (jhenstridge) [inline ci/snapcraft.sh]: Would this be simpler as a single part?  Something like:

```
parts:
  rustup:
    plugin: nil
    build-attributes: [no-patchelf]
    source: inputs
    override-build: |
      mkdir $SNAPCRAFT_PART_INSTALL/bin
      cp rustup-init $SNAPCRAFT_PART_INSTALL/bin/rustup
      ...
```

I believe that should have the same effect as what you've got.

File: ci/snapcraft.sh
Comment (jhenstridge) [inline ci/snapcraft.sh]: Is there any particular reason for using hard links over symlinks here? I'm curious to know if they get preserved as hard links in the resultiing squashfs file.

If they don't get preserved, then symlinks could give a noticeable reduction in package size.

File: ci/snapcraft.sh
Comment (jhenstridge) [inline ci/snapcraft.sh]: If you are storing a lot of data in this directory, it might be worth switching this to `$SNAP_USER_COMMON`. The `$SNAP_USER_DATA` directory is versioned, and will be copied on an upgrade of the snap.  This is done so that configuration can be rolled back if you revert an upgrade of a snap.

If you're happy to deal with data from potentially newer versions of your snap, then `$SNAP_USER_COMMON` is a better choice.  It will save disk space and make upgrades faster.

**Review by jhenstridge — COMMENTED** [review]
I've left a few drive-by comments after the discussion on the Snapcraft forum.  I've been working on an action to publish snaps here:

https://github.com/jhenstridge/snapcraft-publish-action

I haven't quite locked down the API yet, but it is functional.  As the action depends on a secret (the Snap Store credentials), I'm not sure how easy it will be to test within a pull request.

File: ci/actions-templates/linux-builds-template.yaml
Comment (kinnison) [reply ci/actions-templates/linux-builds-template.yaml]: Yeah, this is actually a bogon from when I was diagnosing something in a fork of my own, thanks for the spot.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: I honestly don't know.  This generator was developed when I was at snapcraft summit in Montréal and since then I've not touched it.  Once I've something which works I'll look at cleaning it up more.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: The hardlinks are there because that's what `rustup` normally does and I wasn't interested in risking things.  I believe squashfs does hard links just fine, so unless the snapcraft settings fail to preserve them we ought to be fine.  If you're worried we can try switching to symlinks once the rest is sorted.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: That sounds very sensible.  `rustup` is pretty stable in terms of the content of `~/.rustup` so I'll switch to `$SNAP_USER_COMMON`

**Comment by kinnison** [comment]
@jhenstridge Thank you for your input and review -- I'll look at the publish action you've provided and I can probably test it by adding the secret to my own repo fork and trying it there to ensure it behaves OK.  I'll try and do that soon.

**Comment by kinnison** [comment]
@jhenstridge I think that ought to do the trick, obviously I'm tied to a particular commit of your repo for now.  I tested it over at my own fork and it seemed to do the right thing, so now I need to wait for you to finalise a v1 of the publish action I guess, and then decide on the final comments you made on the snapcraft yaml file.

File: ci/snapcraft.sh
Comment (jhenstridge) [reply ci/snapcraft.sh]: I know squashfs can handle them.  I was more curious about whether snapcraft itself was preserving them as the data is copied from the part's install dir, to the stage dir, and finally to the prime dir.

This may not even be a problem in the end, so I agree that investigating this shouldn't be a blocker for landing this PR.

File: ci/actions-templates/linux-builds-template.yaml
Comment (jhenstridge) [inline ci/actions-templates/linux-builds-template.yaml:188]: This should just be:

    release: edge,beta

I don't think Github actions supports repeating input parameters for actions: they're passed to the action as environment variables rather than anything more structured.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: :+1:

File: ci/actions-templates/linux-builds-template.yaml
Comment (kinnison) [reply ci/actions-templates/linux-builds-template.yaml:188]: Those comments are directives to a script which generates our real workflow files, on PRs, there'll be no `release`, on master builds it'll say `edge` and on stable builds `beta`

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: I can confirm this cleanup works, thanks James.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: I've checked and the snap I end up with on x86_64 linux is 2.7M which is definitely not a multiple of the binary size, so either squashfs has noticed they're all the same and compressed well, or else the hardlinks worked :D

**Comment by kinnison** [comment]
I believe everything is right now, and I've added the token to the repo's secrets ready for post-merge use.  If noone objects, I'll merge this soon and we can see how that goes.

File: .github/workflows/linux-builds-on-master.yaml
Comment (rbtcollins) [inline .github/workflows/linux-builds-on-master.yaml:31]: Should these stay skip-pr?

File: .github/workflows/linux-builds-on-master.yaml
Comment (rbtcollins) [inline .github/workflows/linux-builds-on-master.yaml]: This is an individual developers snap action; I love James to bits, but if he leaves or gets hit by a bus this is going to break. Does Canonical not have a snap team account for this action to live under?

File: ci/snapcraft.sh
Comment (rbtcollins) [inline ci/snapcraft.sh]: suggest -u -f and -o pipefail as well. See https://sipb.mit.edu/doc/safe-shell/ for a longer explanation than just me saying 'its a good idea'.

**Review by rbtcollins — CHANGES_REQUESTED** [review]
One question, one requested change

File: .github/workflows/linux-builds-on-master.yaml
Comment (kinnison) [reply .github/workflows/linux-builds-on-master.yaml:31]: The intent here was PRs only test building the snap, so we check the native arch and one cross arch to make sure I've not fluffed anything.  then master builds more of them because it uploads to the edge channel, and stable uploads all of them to the beta channel.  Does that make sense to you?

File: .github/workflows/linux-builds-on-master.yaml
Comment (kinnison) [reply .github/workflows/linux-builds-on-master.yaml]: Currently they don't.  The eventual goal for `rust-lang` itself is to mirror any actions we depend on which aren't github official ones or actions-rs I think.  For now I also trust James won't go bad, and once I know what the overall plan is, we can see about mirroring those actions or whatever is needed if Canonical don't do official actions in the snapcraft group or somesuch.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Hmm, -u will render the checks at the top redundant or else they'll need rejiggering to cope if they're set to empty strings.  I guess -f would be harmless since there's no * or ? in the script anyway, so I'll add that to future-proof things.  This is a 'sh' so on the builders it'll end up with dash so -o pipefail isn't available according to that link.  I'll add -f and note that this passes shellcheck in case that mollifies you about the others.

File: .github/workflows/linux-builds-on-master.yaml
Comment (rbtcollins) [reply .github/workflows/linux-builds-on-master.yaml:31]: So master = edge, and stable == releases == beta, and actual stable channel is manual?

File: .github/workflows/linux-builds-on-master.yaml
Comment (rbtcollins) [reply .github/workflows/linux-builds-on-master.yaml]: Yeah, I have zero concerns about James going bad :) Could we at least ask if they are willing to make an actions repo in snapcraft/ ? Or has that been discussed and decided against?

File: ci/snapcraft.sh
Comment (rbtcollins) [inline ci/snapcraft.sh]: Its things like this that pipefail is for; you can use mispipe from moreutils if you want to stay with POSIX sh, or switch to bash shell.

File: ci/snapcraft.sh
Comment (rbtcollins) [inline ci/snapcraft.sh]: fixing this for undefined safety is easy:
```
if [ -z "${DO_SNAP:-}" ]; then
```

**Comment by rbtcollins** [comment]
Is the shell script shellchecked automatically?

File: .github/workflows/linux-builds-on-master.yaml
Comment (kinnison) [reply .github/workflows/linux-builds-on-master.yaml:31]: That is indeed the current plan.

File: .github/workflows/linux-builds-on-master.yaml
Comment (kinnison) [reply .github/workflows/linux-builds-on-master.yaml]: I believe James is seeking futher proof that his actions are good, before he persues officialdom.

Comment (kinnison) [reply .github/workflows/linux-builds-on-master.yaml]: I have been discussing the work he did on https://forum.snapcraft.io/t/call-for-testing-github-action-for-snapcrtaft/14930 but as yet he's not indicated an intent to carry it forward further.

Comment (kinnison) [reply .github/workflows/linux-builds-on-master.yaml]: /cc @jhenstridge

**Comment by kinnison** [comment]
@rbtcollins Yes, shellcheck is run over all our shell scripts as part of the CI.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Of course.  I'm so dozy when it comes to shell :D

Comment (kinnison) [reply ci/snapcraft.sh]: Done this where appropriate, thanks Robert.

File: ci/snapcraft.sh
Comment (kinnison) [reply ci/snapcraft.sh]: Aha, yeah I see.  I'll switch to bash and enable pipefail.

File: .github/workflows/linux-builds-on-master.yaml
Comment (jhenstridge) [reply .github/workflows/linux-builds-on-master.yaml]: I have brought up moving these actions to the `snapcore` github organisation, both to make it easier to share maintenance and to make updating them a bit structured.  These actions started as a holiday project to learn about this part of the GH Actions system, so it hasn't been a priority during the week.

You are right that there are security concerns about trusting an action that I solely control. The publish action in particular is in a prime position to silently exfiltrate the store credentials of any project that uses it.  The build action could similarly modify the project before building it: a little more involved and likely to be noticed in the logs, but still a possibility.

Actions provided by `snapcore` could also do all this, but as it is the same org that publishes Snapcraft, it is also in a position to do all of this.

I'll try and move forward the process of moving the projects over.

**Comment by kinnison** [comment]
@rbtcollins I believe your question and change were resolved.  I've updated to use the now official snapcore actions.  I think we're likely good to merge assuming the CI remains green.  Opinions?

**Comment by kinnison** [comment]
I have decided to go ahead and merge this because we need to work through any problems long before we make a release, and we can only do that once it's trying to push to snapstore etc.

**Pull request merged**
